source/common/http/filter_chain_helper.h

```diff
@@ -108,11 +108,13 @@ class FilterChainHelper : Logger::Loggable<Logger::Id::config> {
         proto_config, factory_context_.messageValidationVisitor(), *factory);
     Http::FilterFactoryCb callback =
         factory->createFilterFactoryFromProto(*message, stats_prefix_, factory_context_);
-    dependency_manager.registerFilter(factory->name(), *factory->dependencies());
-    bool is_terminal = factory->isTerminalFilterByProto(*message, server_context_);
-    Config::Utility::validateTerminalFilters(proto_config.name(), factory->name(),
-                                             filter_chain_type, is_terminal,
-                                             last_filter_in_current_config);
+    if (!last_filter_in_current_config || validate_last_filter) {
```

Comment:
What's the story with this new conditional? Is this part of the mysterious router code problems you hit?

Reply:
Was just about to ask the same thing. :) Maybe some comments?

Reply:
haha, this is an artifact of how upstream filters are created.  I already had a TODO in the main PR to clean it up but I'll add an explanation and a TODO here as well.